lib/HLSL/DxilPreparePasses.cpp

```diff
@@ -1023,3 +1028,120 @@ ModulePass *llvm::createDxilValidateWaveSensitivityPass() {
+            Loop *BreakLoop = LInfo->getLoopFor(BB);
+            if (BreakLoop && BreakLoop->contains(WaveBB))
+              SensitiveBB.insert(BB);
+          }
+        }
+      }
+      // TODO: hit the brakes if we've left any loop that might contain WaveCI
+      WorkList.append(I->user_begin(), I->user_end());
+    }
+  }
+}
+
+class RevertWavelessBreaks : public FunctionPass {
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit RevertWavelessBreaks() : FunctionPass(ID) {}
+  const char *getPassName() const override { return "DXIL Unconditionalize breaks"; }
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<LoopInfoWrapperPass>();
+  }
+
+  LoopInfo *LInfo;
+
+  bool runOnFunction(Function &F) override {
+    if (F.isDeclaration())
+      return false;
+    // Only check ps and lib profile.
+    Module *M = F.getEntryBlock().getModule();
+    DxilModule &DM = M->GetDxilModule();
+    const ShaderModel *pSM = DM.GetShaderModel();
+    // If we're not a wave-capable stage, nothing needs to be done
+    if (!pSM->IsPS() && !pSM->IsLib())
+      return false;
+
+    LInfo = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+    // For each wave operation, collect the blocks sensitive to it
+    SmallPtrSet<BasicBlock *, 16> SensitiveBBs;
+    for (Function &IF : M->functions()) {
+      if (&IF == &F || !IF.getNumUses() || !IF.isDeclaration() || !hlsl::OP::IsDxilOpFunc(&IF))
+        continue;
+
+      for (User *U : IF.users())
+        if (CallInst *CI = dyn_cast<CallInst>(U)) {
+          DXIL::OpCode opcode = hlsl::OP::GetDxilOpFuncCallInst(CI);
+          if (OP::IsDxilOpWave(opcode))
+            CollectSensitiveBlocks(LInfo, CI, SensitiveBBs);
+        }
+    }
+
+    bool Changed = false;
+    // Revert artificially conditional breaks in blocks not included in SensitiveBBs
+    for (auto &BB : F) {
+      if (!SensitiveBBs.count(&BB)) {
+        if (BranchInst *BI = dyn_cast<BranchInst>(BB.getTerminator())) {
+          if (BI->isConditional()) {
+            Instruction *Cond = dyn_cast<Instruction>(BI->getCondition());
+            LoadInst *LI = Cond?dyn_cast<LoadInst>(Cond->getOperand(0)):nullptr;
+            if (Cond && isa<ICmpInst>(Cond) && LI && LI->isVolatile()) {
+              // Make branch conditional always true and erase the conditional preds
+              Constant *C = ConstantInt::get(Type::getInt1Ty(BI->getContext()), 1);
+              BI->setCondition(C);
+              Cond->eraseFromParent();
+              LI->eraseFromParent();
+              Changed = true;
+            }
+          }
+        }
+      }
+    }
+    return Changed;
+  }
+};
+
+}
+
+char RevertWavelessBreaks::ID = 0;
+
+INITIALIZE_PASS_BEGIN(RevertWavelessBreaks, "hlsl-waveless", "HLSL Remove false condition", false, false)
+INITIALIZE_PASS_DEPENDENCY(LoopInfoWrapperPass)
+INITIALIZE_PASS_END(RevertWavelessBreaks, "hlsl-waveless", "HLSL Remove false condition", false, false)
```

Comment:
I don't like any of these names, but I'm too tired to think of anything better just now. I'm open to suggestions!